python/pyarrow/tensorflow/plasma_op.cc

```diff
@@ -0,0 +1,381 @@
+      // http://docs.nvidia.com/cuda/cuda-driver-api/group__CUDA__MEM.html#group__CUDA__MEM_1gf0a9fe11544326dabd743b7aa6b54223
+      CHECK(stream_executor->HostMemoryRegister(static_cast<void*>(data),
+                                                static_cast<uint64>(total_bytes)));
+
+      {
+        mutex_lock l(d2h_stream_mu);
+        if (d2h_stream == nullptr) {
+          d2h_stream = new Stream(stream_executor);
+          CHECK(d2h_stream->Init().ok());
+        }
+      }
+
+      // Needed to make sure the input buffers have been computed.
+      // NOTE(ekl): this is unnecessary when the op is behind a NCCL allreduce already
+      CHECK(d2h_stream->ThenWaitFor(orig_stream).ok());
+
+      for (int i = 0; i < num_tensors; ++i) {
+        const auto& input_tensor = context->input(i);
+        auto input_buffer = const_cast<char*>(input_tensor.tensor_data().data());
+        perftools::gputools::DeviceMemoryBase wrapped_src(
+            static_cast<void*>(input_buffer));
+        const bool success =
+            d2h_stream
+                ->ThenMemcpy(static_cast<void*>(data + offsets[i]), wrapped_src,
+                             static_cast<uint64>(offsets[i + 1] - offsets[i]))
+                .ok();
+        OP_REQUIRES_ASYNC(context, success,
+                          errors::Internal("D2H memcpy failed to be enqueued."), done);
+      }
+      context->device()->tensorflow_gpu_device_info()->event_mgr->ThenExecute(
+          d2h_stream, std::move(wrapped_callback));
+#endif
+    }
+  }
+
+ private:
+  std::string plasma_store_socket_name_;
+  std::string plasma_manager_socket_name_;
+
+  mutex mu_;
+  bool connected_ = false;
+  plasma::PlasmaClient client_ GUARDED_BY(mu_);
+};
+
+static Stream* h2d_stream = nullptr;
+static mutex h2d_stream_mu;
+
+// Get:  plasma -> tf.Tensor.
+template <typename Device>
+class PlasmaToTensorOp : public AsyncOpKernel {
+ public:
+  explicit PlasmaToTensorOp(OpKernelConstruction* context) : AsyncOpKernel(context) {
+    OP_REQUIRES_OK(context, context->GetAttr("plasma_store_socket_name",
+                                             &plasma_store_socket_name_));
+    OP_REQUIRES_OK(context, context->GetAttr("plasma_manager_socket_name",
+                                             &plasma_manager_socket_name_));
+    mutex_lock lock(mu_);
+    if (!connected_) {
+      VLOG(1) << "Connecting to Plasma...";
+      ARROW_CHECK_OK(client_.Connect(plasma_store_socket_name_,
+                                     plasma_manager_socket_name_,
+                                     plasma::kPlasmaDefaultReleaseDelay));
+      VLOG(1) << "Connected!";
+      connected_ = true;
+    }
+  }
+
+  ~PlasmaToTensorOp() override {
+    {
+      mutex_lock lock(mu_);
+      ARROW_CHECK_OK(client_.Disconnect());
+      connected_ = false;
+    }
+    {
+      mutex_lock lock(h2d_stream_mu);
+      if (h2d_stream != nullptr) {
+        delete h2d_stream;
+      }
+    }
+  }
```

Comment:
There is a sense of code duplication in the constructor and destructor here vs. the other kernel above. Any way to enable code reuse?